packages/kbn-esql-validation-autocomplete/src/definitions/types.ts

```diff
@@ -8,81 +8,90 @@
 
 import type { ESQLCommand, ESQLCommandOption, ESQLFunction, ESQLMessage } from '@kbn/esql-ast';
 
-// Currently, partial of the full list
-// https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java
-export const supportedFieldTypes = [
+/**
+ * All supported field types in ES|QL. This is all the types
+ * that can come back in the table from a query.
+ */
+export const fieldTypes = [
+  'boolean',
+  'date',
   'double',
-  'unsigned_long',
-  'long',
+  'ip',
+  'keyword',
   'integer',
-  'counter_integer',
-  'counter_long',
-  'counter_double',
-  'date',
-  'date_period',
+  'long',
   'text',
-  'keyword',
-  'boolean',
-  'ip',
+  'unsigned_long',
+  'version',
   'cartesian_point',
   'cartesian_shape',
   'geo_point',
   'geo_shape',
-  'version',
+  'counter_integer',
+  'counter_long',
+  'counter_double',
+  'unsupported',
+] as const;
+
+export type FieldType = (typeof fieldTypes)[number];
+
+export const isFieldType = (type: string | FunctionParameterType): type is FieldType =>
+  fieldTypes.includes(type as FieldType);
+
+/**
+ * This is the list of all data types that are supported in ES|QL.
+ *
+ * Not all of these can be used as field types. Some can only be literals,
+ * others may be the value of a field, but cannot be used in the index mapping.
+ *
+ * This is a partial list. The full list is here and we may need to expand this type as
+ * the capabilities of the client-side engines grow.
+ * https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java
+ */
+export const dataTypes = [
+  ...fieldTypes,
+  'null',
+  'time_literal', // @TODO consider merging time_literal with time_duration
+  'time_duration',
+  'date_period',
```

Comment:
It may make sense to merge all these three, but maybe as a follow-on

Reply:
From @bpintea on `time_duration` and `date_period`

> They cover different time spans and map internally to Java's java.time.Duration and .Period, which cover respectively units until hour and from day. There's no function we have that only covers one of the spans, since these are inter-convertible in the time space.
Thinking about it, we should have probably built a data type that abstracts this internal detail.. but this is how it was implemented early on. (We might want to revisit it, but that's not a priority, I think.)

From @nik9000 on `time_duration` and `time_literal`

> It looks like timespan literals are the way to write a constant time_duration. and time_durations can only be constants at the moment. so they aren't the same, but sure are about the same

So, from the perspective of users in the editor, these may all be functionally equivalent.